[nnc] Added a cache to use singleton instances of PytorchLLVMJIT for every triple,cpu,attrs combination

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66218
* __->__ #66217
* #66216

Every instance of `PytorchLLVMJIT` object takes up `~0.8MB` of memory. In Static Runtime, where we have multiple instances of `LLVMCodeGen` each having its own instance of `PytorchLLVMJIT`, this results in taking up a lot of memory. 

This change creates one instance of `PytorchLLVMJIT` for every unique combination of `triple, cpu, attrs`. This PR also adds a cache to create, manage, and retrieve the instances of `PytorchLLVMJIT`. 

One consequence of using a single instance of `PytorchLLVMJIT` is that all the functions that are added in it should have unique names. So, added the functionality to save a list of function names used so far and generate unique function names (w.r.t. those observed so far).

Differential Revision: [D31445797](https://our.internmc.facebook.com/intern/diff/D31445797)

navahgar committed:
[nnc] Added a cache to use singleton instances of PytorchLLVMJIT for every triple,cpu,attrs combination

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
Overall looks good, only a couple of questions below.

navahgar committed:
Update on "[nnc] Added a cache to use singleton instances of PytorchLLVMJIT for every triple,cpu,attrs combination"


Every instance of `PytorchLLVMJIT` object takes up `~0.8MB` of memory. In Static Runtime, where we have multiple instances of `LLVMCodeGen` each having its own instance of `PytorchLLVMJIT`, this results in taking up a lot of memory. 

This change creates one instance of `PytorchLLVMJIT` for every unique combination of `triple, cpu, attrs`. This PR also adds a cache to create, manage, and retrieve the instances of `PytorchLLVMJIT`. 

One consequence of using a single instance of `PytorchLLVMJIT` is that all the functions that are added in it should have unique names. So, added the functionality to save a list of function names used so far and generate unique function names (w.r.t. those observed so far).

Differential Revision: [D31445797](https://our.internmc.facebook.com/intern/diff/D31445797)

[ghstack-poisoned]
